lib/elixir/lib/kernel/typespec.ex

```diff
@@ -598,28 +616,26 @@ defmodule Kernel.Typespec do
 
         # This may be generating an invalid typespec but we need to generate it
         # to avoid breaking existing code that was valid but only broke dialyzer
-        {left, state} = typespec(var, [elem(var, 0) | vars], caller, state)
         {right, state} = typespec(expr, vars, caller, state)
-        {{:ann_type, line(meta), [left, right]}, state}
+        {{:ann_type, line(meta), [{:var, line(var_meta), var_name}, right]}, state}
 
       {right, state} ->
-        {left, state} = typespec(var, [var_name | vars], caller, state)
-        {{:ann_type, line(meta), [left, right]}, state}
+        {{:ann_type, line(meta), [{:var, line(var_meta), var_name}, right]}, state}
     end
   end
 
-  defp typespec({:::, meta, [var, expr]} = ann_type, vars, caller, state) do
+  defp typespec({:::, meta, [left, right]} = expr, vars, caller, state) do
     message =
       "invalid type annotation. When using the | operator to represent the union of types, " <>
-        "make sure to wrap type annotations in parentheses: #{Macro.to_string(ann_type)}"
+        "make sure to wrap type annotations in parentheses: #{Macro.to_string(expr)}"
 
     # TODO: make this an error in elixir 2.0 and remove code below
     :elixir_errors.warn(caller.line, caller.file, message)
 
     # This may be generating an invalid typespec but we need to generate it
     # to avoid breaking existing code that was valid but only broke dialyzer
-    {left, state} = typespec(var, [elem(var, 0) | vars], caller, state)
-    {right, state} = typespec(expr, vars, caller, state)
+    {left, state} = typespec(left, vars, caller, state)
```

Comment:
This clause is meant for invalid type annotations due to the operator precenden between the type operator `::` and the pipe operator `|`. For example, `@type my_type :: pid | node | my_ann_type :: atom`.

For that reason, I think `elem(var, 0)` makes no sense because it's going to be `:|`